Lazyload iframes

domfarolino opened the pull request:
Fixes https://github.com/whatwg/html/issues/5496. 

- [x] At least two implementers are interested (and none opposed):
   * Chromium
   * Firefox
   * Safari
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   - Tentative tests can be found at https://github.com/web-platform-tests/wpt/tree/master/html/semantics/embedded-content/the-iframe-element. Here's what we still need:
   - [x] Test that the a deferred iframe is fetched with the referrer policy computed at load-deferral time
   - [x] Test that a deferred iframe is fetched with a URL that was parsed against
   - [x] Test generic lazy load behavior (i.e., below-viewport `loading=lazy` iframes do not load)
   - [x] Test generic lazy load eager behavior
   - [x] Test loading=lazy on srcdoc iframes
   - [x] Test that below-viewport `loading=lazy` iframes can be lazy loaded multiple times, and that a document in an iframe isn't unloaded until the second time the `ready to be lazy loaded` flag is loaded
   - [x] From discussion with @annevk, test that setting `loading=eager` and then immediately removing the attribute still continues the load (we need this for images too)
   - [x] Test that setting `loading=eager` on a deferred iframe correctly sets the `ready to be lazy loaded flag` and continues the load
   - [x] Test that in-viewport loading=lazy iframes do not block the outer window load event
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: N/A
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1622090
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=196698

Aside: One thing I noticed about https://github.com/whatwg/html/pull/5510 is that I think it didn't fully fix https://github.com/whatwg/html/issues/5236. I think it's still weird that we're waiting in parallel for the `ready to be lazy loaded` bool to be mutated on the main thread, and then we queue a task from in-parallel to continue. IMO one of the big wins of using an IntersectionObserver is that we can use the callback to just synchronously finish the loading right from there. I'll start another thread with my latest proposed design to fix this, but for now, the iframe spec behaves similar to images.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/5579/embedded-content.html" title="Last updated on Jul 1, 2020, 4:44 PM UTC (d781928)">/embedded-content.html</a>  ( <a href="https://whatpr.org/html/5579/4db63e0...d781928/embedded-content.html" title="Last updated on Jul 1, 2020, 4:44 PM UTC (d781928)">diff</a> )
<a href="https://whatpr.org/html/5579/iframe-embed-object.html" title="Last updated on Jul 1, 2020, 4:44 PM UTC (d781928)">/iframe-embed-object.html</a>  ( <a href="https://whatpr.org/html/5579/4db63e0...d781928/iframe-embed-object.html" title="Last updated on Jul 1, 2020, 4:44 PM UTC (d781928)">diff</a> )
<a href="https://whatpr.org/html/5579/images.html" title="Last updated on Jul 1, 2020, 4:44 PM UTC (d781928)">/images.html</a>  ( <a href="https://whatpr.org/html/5579/4db63e0...d781928/images.html" title="Last updated on Jul 1, 2020, 4:44 PM UTC (d781928)">diff</a> )
<a href="https://whatpr.org/html/5579/indices.html" title="Last updated on Jul 1, 2020, 4:44 PM UTC (d781928)">/indices.html</a>  ( <a href="https://whatpr.org/html/5579/4db63e0...d781928/indices.html" title="Last updated on Jul 1, 2020, 4:44 PM UTC (d781928)">diff</a> )
<a href="https://whatpr.org/html/5579/urls-and-fetching.html" title="Last updated on Jul 1, 2020, 4:44 PM UTC (d781928)">/urls-and-fetching.html</a>  ( <a href="https://whatpr.org/html/5579/4db63e0...d781928/urls-and-fetching.html" title="Last updated on Jul 1, 2020, 4:44 PM UTC (d781928)">diff</a> )

annevk commented:
As specified it doesn't account for `srcdoc`, meaning that if `srcdoc` specifies further subresources, those all end up delaying the `load` event. That seems less than ideal. I think a primary use case for this feature is to make the `load` event fire faster. (It seems fine to not account for when neither `srcdoc` nor `src` is specified.)

The other problem I foresee here is that navigation obtains certain state at which point we run into nondeterminism in the sense that it matters when the user scrolls the page. This already exists to some extent, e.g., with the `sandbox` attribute, and this will enhance that.

cc @clelland

domenic commented:
@domfarolino asked me to look at this to get a second set of eyes on the concern

> The other problem I foresee here is that navigation obtains certain state at which point we run into nondeterminism in the sense that it matters when the user scrolls the page. This already exists to some extent, e.g., with the sandbox attribute, and this will enhance that.

I tried to get an exhaustive list of all the mutable state that affects navigation. They are:

- Handled in the "otherwise steps"
  - container `src=""` attribute
  - embedder document base URL (affects URL parsing)
  - ancestor tree active document URLs (to avoid recursion)
  - container `referrerpolicy=""` attribute
- Handled in the "navigate" algorithm
  - sandboxing flags of the embedder and of the container (impacts both "allowed to navigate" and the creation sandboxing flags of the new document)
- Handled in "process a navigate fetch"
  - CSP of embedder

Given how non-uniformly these are handled right now, my intuition is that the cleanest thing to do is to proceed as this PR does, by making minimal changes to the "navigate" algorithm and instead doing the deferral in the "otherwise steps". That leaves us better equipped to clean these up in the manner suggested in #4926 later.

That is, if we go mucking around inside the navigation algorithm (by e.g. moving the deferral until right before "process a navigate fetch"), that will make future refactoring into an up-front policy-container harder. (Or at the very least we'd have to move the deferral back up into the otherwise steps.) And we'd end up having to duplicate a lot of the checks in "navigate" after the deferral runs its course, since they are important for security.

So if we go with this PR as-is, this would give us the following behavior for images vs. iframes. Here I used "loading-started" to refer to the beginning of the relevant loading algorithm (usually right after the element gets inserted into the document), and "loading-committed" to refer to the point after which we're about to fetch from the network (e.g., once the user has scrolled down enough).

- `<img>`:
  - Snapshotted at loading-started: embedder document base URL, `crossorigin=""`
  - Causes a re-fetch: `src=""`, `referrerpolicy=""`
  - Consulted after loading-committed: embedder CSP
- `<iframe>`:
  - Snapshotted at loading-started: embedder document base URL, `referrerpolicy=""`
  - Causes a re-fetch: `src=""`
  - Consulted after loading-committed: embedder CSP, embedder sandbox, container `sandbox=""`

This seems like an acceptable intermediate state for me. If I were to change anything, it would be to find out if we can uniformize the behavior of `referrerpolicy=""`. And maybe try to make the container `sandbox=""` behave like `referrerpolicy=""`, but that's a stretch.

Now, what is the ultimate desired state? We should probably discuss that in another issue, perhaps #4926 or perhaps a dedicated one. #4926 gestures at moving everything into "snapshotted at loading-started". But getting there will be pretty difficult, especially given the extensive dependency on browsing contexts in CSP and sandboxing. And the discussion might impact other fetches, such as `<img>` or perhaps even `<a>`.

domfarolino commented:
Thanks a ton for the detailed feedback @domenic! I have a couple of corrections:

> This seems like an acceptable intermediate state for me. If I were to change anything, it would be to find out if we can uniformize the behavior of referrerpolicy="". And maybe try to make the container sandbox="" behave like referrerpolicy="", but that's a stretch.

I'll go ahead and edit your comment, but actually the image loading process is interesting: We arguably "snapshot" every property of the request before **load-committed**, but both `referrerpolicy` and `crossorigin` mutations cause-refetches, along with all other [relevant mutations](https://html.spec.whatwg.org/multipage/images.html#relevant-mutations).

Given this, I think uniformizing these attributes would require introducing something like a relevant mutations equivalent for the iframe element, which would basically involve:
 - Creating this definition for iframes
 - Including things like [iframe-7](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#the-iframe-element:the-iframe-element-7), [iframe-9](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#the-iframe-element:the-iframe-element-9), and [iframe-10](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#the-iframe-element:the-iframe-element-10) in this list
 - Also mentioning `referrerpolicy` mutations in this list

But yeah I'm fine with that being done separately, if it is desirable.

Given all this, I think this is ready for review!

domfarolino commented on source:
```diff
@@ -29803,6 +29806,21 @@ interface <dfn>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
+
+     <ol>
+      <li><p>Set the <code>iframe</code>'s <span>lazy load resumption steps</span> to the rest of
+      this algorithm.</p></li>
```
Can you clarify the example? Are you saying to use a `<dl class="switch">`?

domenic replied:
FWIW I also found this a bit confusing, but it's the same as for image, so it might be better to keep things consistent.

If I were to refactor these, I would just go with a bit of repetition:

> Set the iframe's lazy load resumption steps to an algorithm that runs start intersection-observing a lazy loading element for the iframe element.

Same for https://html.spec.whatwg.org/#updating-the-image-data:lazy-load-resumption-steps.

domenic replied:
Oh shoot. I think I misunderstood what you meant. You mean step 5? And for image, you mean steps 23-25?

In that case then yeah this is pretty bad. Two fixes:

1. A bit of repetition + refactoring. "Set the iframe's lazy load resumption steps to an algorithm that navigates the element's nested browsing context to resource." And refactor steps 23-25 of img to something like a "Fetch the image" algorithm, so you can similarly duplicate it as the new step 23 and also as the steps.

2. What, I believe, Simon was suggesting: add a step label, e.g. "5. _Navigate_:" and "23. _Fetch_: ...", and then say "Set the iframe's lazy load resumption steps to the steps to the portion of this algorithm starting at _Navigate_", or something of that sort.

domfarolino replied:
Thanks, got it! I went with the "label" route since it seemed pretty minimal compared to introducing new algorithms

zcorpan replied:
Indeed (2) was what I meant. Changes look good.

domfarolino commented:
Requesting review from @domenic as well given his feedback, and the current state of the PR.

domenic commented:
> but both referrerpolicy and crossorigin mutations cause-refetches, along with all other relevant mutations.

That isn't quite true, right? Because of the image cache, which is keyed by `crossorigin=""`, changing `crossorigin=""` will not cause a re-fetch.

domenic commented on source:
```diff
@@ -29803,6 +29807,21 @@ interface <dfn>HTMLIFrameElement</dfn> : <span>HTMLElement</span> {
+    <p>If the <span>will lazy load element steps</span> given the <code>iframe</code> return true,
+    then:</p>
+
+     <ol>
```
Indentation looks off here.

domfarolino commented:
The main unresolved question I think we have is should `loading=lazy` have an effect on `<iframe srcdoc="...">`? Currently I believe Chrome does not support this. I'm indifferent on what the spec should say / do.

annevk commented:
I think it's a little more complicated. A `srcdoc` document can include subresources and those will block the `load` event of that document which in turn blocks the `load` event of the parent document. Not blocking the `load` event of the parent document seems very useful and a core aspect of "lazy loading".

domenic commented:
So I think this was discussed a bit in IRC, but it'd be good to record the conclusion in the thread. As I see it we have three options:

1. The PR as-is, with the understanding that for now lazy doesn't do anything on srcdoc, but we'll add it in a followup.

2. No lazy srcdocs: in this case the PR should add an authoring conformance requirement prohibiting the combination.

3. Support lazy srcdocs now: in this case the PR needs updating to support that.

What's the plan?

domfarolino commented:
I believe the plan is (3), to support srcdoc in this PR and test it. I'll try and do at least one of those today or tomorrow.

domfarolino commented:
The latest commit(s) add `loading=lazy` support for `srcdoc` iframes, PTAL @emilio and others! It does so by adding a little duplication between the `srcdoc` steps and the "otherwise" steps, but it's pretty minor so I think it's OK.

I also fixed what I consider a small bug that I added in the _otherwise_ frame steps. When we run the `will lazy load element steps`, we assumed the element was an `iframe`, however we share these steps with the frameset tag. I fixed this, so now we check to see if the element is an `iframe` before invoking the `will lazy load image steps`, and don't do so otherwise.

domfarolino commented:
Can someone sanity check me here? I want to make sure that `loading=lazy` iframes are not accidentally blocking their containing document's load event when we're waiting for their resumption steps to be called. When we're waiting, I don't believe the [potentially delays the load event](https://html.spec.whatwg.org/multipage/iframe-embed-object.html#potentially-delays-the-load-event) conditions are met, which means we're good, but I'd like someone to verify. My rationale for the _potentially delays the load event_ conditions not being met:

 - > element's nested browsing context's active document is not ready for post-load tasks.
    - When the iframe creates a new nested browsing context, we create a new browsing context, and [the document created here](https://html.spec.whatwg.org/multipage/browsers.html#creating-browsing-contexts:concept-document-feature-policy) is immediately ready for post-load tasks, so this condition is not met
 - > Anything is delaying the load event of element's nested browsing context's active document.
    - Nope
 - > element's nested browsing context is in the delaying load events mode.
    - Not yet, but it will be later in the Navigate algorithm

So I think we're good. Buttttt, this brings me to a difference I spotted between iframe and image lazy loading. For image lazy loading, even if the request is resumed before the window load event is fired, the request doesn't delay the window load event. This is due to [#updating-the-image-data-step21](https://html.spec.whatwg.org/multipage/images.html#updating-the-image-data:lazy-loading-attribute). However for iframes, even if the load is resumed before the containing document's window load event is fired, then eventually we'll end up in the Navigation algorithm, and the nested browsing context will be put into the `delaying load events mode`, and the other _potentially delays the load event_ conditions may be met as well.

I think one non-invasive way to fix this would be to keep some state that is persistent per-iframe-request that basically says `current navigation was lazy loaded`. This only gets reset once the "otherwise steps" or `srcdoc` navigation steps are re-invoked. We reference this new state / bool in the following way. We change:

> The iframe element _potentially delays the load event_.

...to say...

> Iframe elements whose _current navigation was lazy loaded_ bool is false _potentially delay the load event_.

This gives us a per-request state independent of the `loading` attribute after resumption, just like the image #updating-the-image-data's _delay load event_ boolean. Not sure if this is the best way. Thoughts?

Either way, we'll need a test like so:
 - You have an in-viewport `loading=lazy` iframe
 - The containing document's window load event is delayed by ~1 second or so
 - After first-render the in-viewport iframe request is resumed, and the request is made before the window load event
 - The request takes ~2 seconds or so
 - The outer window load event is fired
 - The iframe eventually finishes load, and its load event is fired (after the outer window one obviously)

annevk commented:
I think you're correct on both counts.

domfarolino commented:
Thanks for checking, please take a look. I've added a commit that introduces the `current navigation was lazy loaded` boolean mentioned above, which is used to influence whether or not an iframe "potentially delays the load event". I'll add a test for this as well, similar to the one we have for images. If we think there's a cleaner way to do this that would require some more shuffling around, I propose filing a bug and taking care of it as a follow-up task.

zcorpan commented on source:
```diff
@@ -25966,8 +25967,9 @@ interface <dfn>HTMLImageElement</dfn> : <span>HTMLElement</span> {
 
-  <p>When the <code data-x="attr-img-loading">loading</code> attribute is set to the <span
-  data-x="attr-loading-eager-state">Eager</span> state, the user agent must run these steps:</p>
+  <p>When the <code data-x="attr-img-loading">loading</code> attribute's value is changed to the
```
```suggestion
  <p>When the <code data-x="attr-img-loading">loading</code> attribute's state is changed to the
```

domfarolino commented:
All tests have landed, PTAL @emilio @annevk

domenic commented:
Alright, this has accumulated quite a set of approvals by now. I'll merge this tomorrow unless anyone else wants to take a final look.
